proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4209,1003 @@ lemma refillUnblockCheck_corres':
+       apply wpsimp
+      apply wpsimp
+     apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+      apply (wpsimp wp: handle_overrun_loop_body_non_zero_refills)
+     apply wpsimp
+    apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+     apply wpsimp
+    apply wpsimp
+   apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+    apply (wpsimp wp: handle_overrun_loop_body_refills_unat_sum_equals_budget)
+   apply wpsimp
+  apply (wpsimp simp: handle_overrun_loop_body_def)
+  apply (rename_tac sc n)
+  apply (subst unat_sub)
+   apply (prop_tac "sc_at (cur_sc s') s'", simp)
+   apply (frule_tac usage=r and s=s' in head_time_buffer_simp)
+   apply (clarsimp simp: sc_at_ppred_def obj_at_def)
+  apply (clarsimp simp: obj_at_def vs_all_heap_simps)
+  apply (frule_tac x="refill_hd sc" in bspec, fastforce)
+  apply (prop_tac "0 < unat r")
+   apply (prop_tac "sc_at (cur_sc s') s'")
+    apply (clarsimp simp: obj_at_def is_sc_obj_def)
+   apply (frule_tac usage=r and s=s' in head_time_buffer_simp)
+   apply (clarsimp simp: sc_at_ppred_def obj_at_def)
+   apply (frule_tac x="refill_hd sc" in bspec, fastforce)
+   apply (fastforce simp: word_le_nat_alt)
+  apply fastforce
+  done
+
+lemma handleOverrunLoop_corres:
+  "usage = usage' \<Longrightarrow>
+   corres (=) (\<lambda>s. sc_at (cur_sc s) s \<and> is_active_sc2 (cur_sc s) s
+                   \<and> pspace_aligned s \<and> pspace_distinct s
+                   \<and> valid_objs s
+                   \<and> pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) (cur_sc s)
+                   \<and> pred_map (\<lambda>cfg. \<forall>refill\<in>set (scrc_refills cfg). 0 < unat (r_amount refill))
+                               (sc_refill_cfgs_of s) (cur_sc s)
+                   \<and> pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) = unat (scrc_budget cfg))
+                               (sc_refill_cfgs_of s) (cur_sc s))
+              (\<lambda>s'. valid_refills' (ksCurSc s') s')
+              (handle_overrun_loop usage)
+              (handleOverrunLoop usage')"
+  apply (rule_tac Q="\<lambda>s'. sc_at' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: sc_at_cross simp: state_relation_def)
+  apply (rule_tac Q="\<lambda>s'. is_active_sc' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: is_active_sc'_cross simp: state_relation_def)
+  apply (clarsimp simp: handle_overrun_loop_def handleOverrunLoop_def runReaderT_def)
+  apply (rule corres_whileLoop_abs; simp?)
+       apply (frule_tac usage=r' in head_time_buffer_equiv; simp?)
+       apply fastforce
+      apply (corressimp corres: handleOverrunLoopBody_corres)
+     apply (wpsimp wp: handle_overrun_loop_body_no_fail)
+     apply (clarsimp simp: vs_all_heap_simps)
+    apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+        apply (rule hoare_weaken_pre)
+         apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+          apply wpsimp
+         apply wpsimp
+        apply wpsimp
```

Comment:
I really can't get `wps` to work. @corlewis do you have any tips?

I did experiment before with a little Eisbach method which does work

```
method wp_proj for proj :: "'a state => _" uses wp simp
   = (rule hoare_weaken_pre, rule hoare_lift_Pf2[where f=proj], wpsimp wp: wp simp: simp, wpsimp+)
```

It unfortunately requires you to write `apply (wp_proj "cur_sc :: det_state => _")` if you'd like to use it, and I can't see how to get rid of the type annotation. Plus, I'm sure it's less general than `wps`, so it's not ideal

Reply:
The `'a state` in the method declaration can be replaced with `_` and this will also work for the `kernel_state`. But still, I'm not that happy with this approach